sdk/storage/Azure.Storage.DataMovement.Blobs/src/BlobContainerClientExtensions.cs

```diff
@@ -0,0 +1,52 @@
+using Azure.Storage.DataMovement;
+using Azure.Storage.DataMovement.Blobs;
+
+namespace Azure.Storage.Blobs
+{
+    /// <summary>
+    /// Data movement extension methods for the <see cref="Azure.Storage.Blobs.BlobContainerClient"/>.
+    /// </summary>
+    public static class BlobContainerClientExtensions
+    {
+        private static Lazy<TransferManager> s_defaultTransferManager = new Lazy<TransferManager>();
+
+        /// <summary>
+        /// Uploads the entire contents of local directory to the blob container.
+        /// </summary>
+        /// <param name="client">The <see cref="BlobContainerClient"/> used for service operations.</param>
+        /// <param name="localDirectoryPath">The full path to the local directory to be uploaded.</param>
+        /// <param name="blobDirectoryPrefix">Optionally specifies the directory prefix to be prepended to all uploaded files.</param>
+        /// <param name="options">Options which control the directory upload.</param>
+        /// <returns>A <see cref="DataTransfer"/> instance which can be used track progress and wait for completion with <see cref="DataTransfer.AwaitCompletion"/>.</returns>
+        public static async Task<DataTransfer> StartUploadDirectoryAsync(this BlobContainerClient client, string localDirectoryPath, string blobDirectoryPrefix = default, BlobContainerClientTransferOptions options = default)
+        {
+            StorageResourceContainer localDirectory = new LocalDirectoryStorageResourceContainer(localDirectoryPath);
+
+            StorageResourceContainer blobDirectory = new BlobDirectoryStorageResourceContainer(client, blobDirectoryPrefix, options?.BlobContainerOptions);
+
+            return await s_defaultTransferManager.Value.StartTransferAsync(localDirectory, blobDirectory, options?.TransferOptions).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Downloads the contents of a blob container.
+        /// </summary>
+        /// <param name="client">The <see cref="BlobContainerClient"/> used for service operations.</param>
+        /// <param name="localDirectoryPath">The full path to the local directory where files will be dowloaded.</param>
+        /// <param name="blobDirectoryPrefix">Optionally restricts the downloaded content to blobs with the specified directory prefix.</param>
+        /// <param name="options">Options which control the container download.</param>
+        /// <returns></returns>
+        public static async Task<DataTransfer> StartDownloadDirectoryAsync(this BlobContainerClient client, string localDirectoryPath, string blobDirectoryPrefix = default, BlobContainerClientTransferOptions options = default)
+        {
+            StorageResourceContainer localDirectory = new LocalDirectoryStorageResourceContainer(localDirectoryPath);
+
+            StorageResourceContainer blobDirectory = new BlobDirectoryStorageResourceContainer(client, blobDirectoryPrefix, options?.BlobContainerOptions);
```

Comment:
if `blobDirectoryPrefix == default` should this be making a `BlobStorageResourceContainer` instead of a root directory storage resource container instead?

Reply:
Maybe applies to the upload as well?

Reply:
Fixed. However, it does beg the question do we need two different types for virtual directories and containers? Could the `BlobStorageResourceContainer` have an optional directory prefix, where `null` indicates the base of the container?

Reply:
Probably when we go to do `UploadContainer` and stuff we can add those. Since those might come with whatever attributes it could be made with. For now this looks good.